lib/Doctrine/ORM/UnitOfWork.php

```diff
@@ -2531,6 +2568,86 @@ public function getSingleIdentifierValue($entity)
     }
 
     /**
+     * @param array  $id
+     * @param string $rootClassName
+     *
+     * @return GhostObjectInterface|object
+     */
+    private function tryGetByIdOrLoadProxy(array $id, string $rootClassName)
+    {
+        if ($fetched = $this->tryGetById($id, $rootClassName)) {
+            return $fetched;
+        }
+
+        $class = $this->em->getClassMetadata($rootClassName);
+
+        if ($class->getSubClasses()) {
+            // can't do this for inheritance trees!
+            // @TODO fetching from the EntityManager feels dirty here
+            return $this->em->find($rootClassName, $id);
+        }
+
+        $sortedId = [];
+
+        foreach ($class->identifier as $idField) {
+            $sortedId[$idField] = $id[$idField];
+        }
+
+        $proxy = $this->em->getProxyFactory()->getProxy($rootClassName, $sortedId);
+
+        $this->registerManaged($proxy, $sortedId, []);
+
+        return $proxy;
+    }
+    /**
+     * @TODO refactor: `getProxy`: no scalars if the identifier is an association, use real value
+     * @TODO note that this mess is recursive, so we need to fix it somewhere else. An identifier
+     *       may be composed by multiple levels of association fetching, where:
+     *       A#id = B(to-one)
+     *       B#id = C(to-one)
+     *       C#id = D(to-one)
+     *       E#id = E(to-one)
+     *       E#id = composite scalar
+     *       that is a mess, but it is the real world scenario
+     * @TODO we need to therefore have a generic utility that, given a scalar identifier, gives us
+     *       a "normalized" identifier. This is pretty much the opposite of what the
+     *       `IdentifierFlattener` does
+     * @TODO note that we also need to sort the identifier fields here
+     *
+     * @TODO Start of the block to be refactored into "make a deep identifier from a flat one"
+     *
+     * @return mixed[]
+     */
+    private function convertFlatIdentifierIntoRealIdentifierFieldValues(
+        ClassMetadata $targetClass,
+        array $flatIdentifier
+    ) {
+
+        $normalizedAssociatedId = [];
+
+        foreach ($targetClass->getDeclaredPropertiesIterator() as $name => $declaredProperty) {
+            if (! \array_key_exists($name, $flatIdentifier)) {
+                continue;
+            }
+
+            if ($declaredProperty instanceof ToOneAssociationMetadata) {
+                $targetIdMetadata = $this->em->getClassMetadata($declaredProperty->getTargetEntity());
+
+                $normalizedAssociatedId[$name] = $this->tryGetByIdOrLoadProxy(
+                    [reset($targetIdMetadata->identifier) => $flatIdentifier[$name]], // @TODO this is where stuff breaks with composite IDs :-(
```

Comment:
This assumes single-field identifiers on associated proxies: extremely dangerous